### ISSUE TITLE
Behaviour of certain Filebeat modules changed with Elasticsearch 8.7.0

### ISSUE THREAD

**Issue by rdner** [elastic/beats#35133]
When our automation updated the Elasticsearch snapshot to 8.7.1 we started having integration test failures.

This can be seen in this PR https://github.com/elastic/beats/pull/34589

To summarise the failures:

There are 6 failures in total, all coming from `mage pythonintegtest` in Filebeat.
My guess, they're all coming from the same [breaking change](https://github.com/elastic/elasticsearch/pull/93179) in Elasticsearch 8.7.0 (see below). This seems like the new stricter JSON parser in Elasticsearch revealed some real bugs that we have in the Filebeat modules/integrations.

### elasticsearch slowlog module: 

Failing files:

 * [test.log](https://github.com/elastic/beats/blob/main/filebeat/module/elasticsearch/slowlog/test/test.log)
 * [auditlog_index_indexing_slowlog.log](https://github.com/elastic/beats/blob/main/filebeat/module/elasticsearch/slowlog/test/auditlog_index_indexing_slowlog.log)

Error message:

> cannot access method/field [slowlog] from a null def reference 

I tried to investigate this and created the issue with the Elasticsearch team where I found out about the breaking change: https://github.com/elastic/elasticsearch/issues/95086 (see for the comments)

TLDR; the module's ingest pipeline is trying to parse invalid JSON, extracted by the grok pattern. There is a trailing `]` in the `source[<multiline_JSON>]` case. It worked before only because Elasticsearch was silently dropping documents not returning an error.

I suspect that the invalid JSON was not generated by the tests intentionally and it's a result of some ingest pipeline bug and a human mistake during creation of `*-expected.json` files.

My guess for the bug would be this grok expression here: 

https://github.com/elastic/beats/blob/ca13ce86cf08d3aae43789650819cac33632cf6e/filebeat/module/elasticsearch/slowlog/ingest/pipeline-json-8.yml#L24-L33

I didn't investigate the rest of the test failures but I suspect they also have something to do with the new stricter JSON parser.

### mysql slowlog module

Failing files:

 * [mariadb-10.3.13.log](https://github.com/elastic/beats/blob/ca13ce86cf08d3aae43789650819cac33632cf6e/filebeat/module/mysql/slowlog/test/mariadb-10.3.13.log)
 * [mysql-ubuntu-8.0.15.log](https://github.com/elastic/beats/blob/ca13ce86cf08d3aae43789650819cac33632cf6e/filebeat/module/mysql/slowlog/test/mysql-ubuntu-8.0.15.log)
 * [percona-ubuntu-8.0.15.log](https://github.com/elastic/beats/blob/ca13ce86cf08d3aae43789650819cac33632cf6e/filebeat/module/mysql/slowlog/test/percona-ubuntu-8.0.15.log)

Error message: 

> where 1 = len({'type_changes': {"root['mysql.thread_id']": {'old_type': <class 'str'>, 'new_type': <class 'list'>, 'old_value': '16', 'new_value': [16, '16']}}})

### iis slowlog module

Failing files:

 * [test-iis-7.5.log](https://github.com/elastic/beats/blob/main/filebeat/module/iis/access/test/test-iis-7.5.log)

Error message: 

> where 3 = len({'type_changes': {"root['source.address']": {'old_type': <class 'str'>, 'new_type': <class 'list'>, 'old_value': '10.1...d.ip'], root['event.type'], root['destination.ip']], 'iterable_item_removed': {"root['event.category'][0]": 'network'}})

**Comment by klacabane** [comment]
Regarding the elasticsearch failures, it looks like the source json that spans multiple lines is treated as an independent log entry routed to pipeline-json while it should be part of the plaintext entry. I'm wondering if this case was set to test invalid format, or if this can actually happen. I've asked in the es slack channel.
If we're testing an invalid format then we should ignore that entry, otherwise we should fix the grok pattern

**Comment by klacabane** [comment]
Looking at the expected documents from the test file that contains multiline json sources, we can see that the grok pattern [stops when it reads a new line](https://github.com/elastic/beats/blob/main/filebeat/module/elasticsearch/slowlog/test/test.log-expected.json#L159), so the grok pattern does not work for this use case. We need to clarify whether this can happen or not.
@dakrone (pinging since you were involved in https://github.com/elastic/elasticsearch/issues/95086) do you know if multiline log entries are possible in slowlogs ?

**Comment by jsoriano** [comment]
Looked a bit into the failures in the mysql module, it looks like some behavioural change in the grok processor. I see two main cases:
* Different types when no type is defined. On these ambiguous cases we can probably set an specific type.
```
'old_value': '8', 'new_value': 8
```
* Repeated values when there was a single value before.
```
'old_value': 'employees-test', 'new_value': ['employees-test', 'employees-test']
```

The second case seems to happen when there are redundant information in logs, and both are parsed. For example here the schema "employees-test" can be obtained from two places:
```
/usr/sbin/mysqld, Version: 10.3.13-MariaDB-1:10.3.13+maria~bionic-log (mariadb.org binary distribution). started with:
Tcp port: 3306  Unix socket: /var/run/mysqld/mysqld.sock
Time                Id Command  Argument
# Time: 190324 16:03:00
# User@Host: root[root] @ localhost []
# Thread_id: 37  Schema: employees-test  QC_hit: No
# Query_time: 2.461578  Lock_time: 0.000196  Rows_sent: 10  Rows_examined: 3145718
# Rows_affected: 0  Bytes_sent: 319
# Tmp_tables: 1  Tmp_disk_tables: 0  Tmp_table_sizes: 4026528
# Full_scan: Yes  Full_join: No  Tmp_table: Yes  Tmp_table_on_disk: No
# Filesort: Yes  Filesort_on_disk: No  Merge_passes: 0  Priority_queue: Yes
use employees-test;
SET timestamp=1553443380;
SELECT last_name, MAX(salary) AS salary FROM employees
    INNER JOIN salaries ON employees.emp_no = salaries.emp_no
    GROUP BY last_name
    ORDER BY salary DESC
    LIMIT 10;
```

Previously only one value was stored, but now it seems that the same value is stored twice. This may be more difficult to fix because I don't know if we can ensure that these values are always redundant, or if we have to rely on only one of them.

On these cases the indexed documents will be probably the same at the end, so this is an issue on testing, that probably doesn't affect customers unless they are using `_source`.

**Comment by jsoriano** [comment]
> Previously only one value was stored, but now it seems that the same value is stored twice.

This seems to be due to this change: https://github.com/elastic/elasticsearch/pull/92586

**Comment by dakrone** [comment]
> @dakrone do you know if multiline log entries are possible in slowlogs ?

I don't think they are currently possible (I tried to reproduce this locally). However, the test seems to be an example of a document from 2018 (version 6.3.0), so I can't say whether at any point in the past it was possible.

**Comment by ebeahan** [comment]
* Fix to address `envoyproxy` failure: https://github.com/elastic/beats/pull/35159
* Issue to track a fix for `mysql_enterprise` failures: https://github.com/elastic/beats/issues/35158

**Comment by lalit-satapathy** [comment]
Added @ishleenk17 for mysql and IIS fixes.

**Comment by ishleenk17** [comment]
The issue with MYSQL and IIS is `not related to Stricter JSON parsing.` There is some change in grok processor behaviour post 8.7.0.

`It reports duplicate values (list) in contrast to a string value before.
`

### For MYSQL:
 [Maria DB Log](https://github.com/elastic/beats/blob/ca13ce86cf08d3aae43789650819cac33632cf6e/filebeat/module/mysql/slowlog/test/mariadb-10.3.13.log)

**Issue**: {"root['mysql.slowlog.schema']": {'`old_type': <class 'str'>, 'new_type': <class 'list'>, 'old_value': 'employees-test', 'new_value': ['employees-test', 'employees-test']}}}`

[MYSQL Ubuntu Logs
](https://github.com/elastic/beats/blob/ca13ce86cf08d3aae43789650819cac33632cf6e/filebeat/module/mysql/slowlog/test/mysql-ubuntu-8.0.15.log)

**Issue**: {"root['mysql.thread_id']": {'`old_type': <class 'str'>, 'new_type': <class 'list'>, 'old_value': '16', 'new_value': [16, '16']`}}}

[Percona Ubuntu Logs
](https://github.com/elastic/beats/blob/ca13ce86cf08d3aae43789650819cac33632cf6e/filebeat/module/mysql/slowlog/test/percona-ubuntu-8.0.15.log)

**Issue**: {"root['mysql.slowlog.schema']": {`'old_type': <class 'str'>, 'new_type': <class 'list'>, 'old_value': 'employees', 'new_value': ['employees', 'employees'`]}}}



### For IIS:

[IIS Logs 7.5
](https://github.com/elastic/beats/blob/main/filebeat/module/iis/access/test/test-iis-7.5.log)

Issue:  {"root['destination.address']": `{'old_type': <class 'str'>, 'new_type': <class 'list'>, 'old_value': '10.100.220.70', 'new_value': ['10.100.220.70', '10.100.220.70']}
`


We are trying to figure out what changed in grok processing ElasticSearch after 8.7.0. One of the related [issue](https://github.com/elastic/elasticsearch/pull/92586) looks to be this one. 
If this is a legitimate processing, then we can simply update the golden file of the expected json.
If not, then could it be a bug in Elasticsearch? 

cc: @leehinman @ruflin @lalit-satapathy @andresrc

**Comment by bhapas** [comment]
>It reports duplicate values (list) in contrast to a string value before. 

An example of what @ishleenk17 mentioned. We found similar issue in one of the integration pipelines - [here](https://github.com/elastic/integrations/pull/5910/files#diff-dae33f24a9ab60fd8007ceea3d2f5d92cfcff6821565c094708734c0fb929fb3L85-L89)

Snippet here

```
  - grok:
      field: cybersixgill.pattern
      patterns:
        - "^\\[%{MD5}\\]"
        - "^\\[%{MD5} OR %{SHA1}\\]"
        - "^\\[%{MD5} OR %{SHA1} OR %{SHA256}\\]"
        - "^\\[%{SHA1}\\]"
        - "^\\[%{SHA1} OR %{SHA256}\\]"
        - "^\\[%{SHA256}\\]"
        - "^\\[%{DATA:threat.indicator.type}:value%{SPACE}=%{SPACE}'%{DATA:_temp_.threatvalue}'\\]"
      pattern_definitions:
        MD5: "%{DATA:threat.indicator.type}:hashes.MD5%{SPACE}=%{SPACE}'%{WORD:threat.indicator.file.hash.md5}'"
        SHA1: "%{DATA:threat.indicator.type}:hashes.'SHA-1'%{SPACE}=%{SPACE}'%{WORD:threat.indicator.file.hash.sha1}'"
        SHA256: "%{DATA:threat.indicator.type}:hashes.'SHA-256'%{SPACE}=%{SPACE}'%{WORD:threat.indicator.file.hash.sha256}'"
```

If we take an example log snippet that caused the issue
```
"pattern":"[file:hashes.MD5 = '31aef6bddfeeb3f519dfe3d5ebe9c2ae' OR file:hashes.'SHA-1' = 'e54ef45c82899dd2b20372cf47958cea94dd80a7' OR file:hashes.'SHA-256' = '2e7e43be1fc3cbefef8d686ce63ceb30456a4a67d555407fb6797e969972945c']"
```
SO the Grok matches on `- "^\\[%{MD5} OR %{SHA1} OR %{SHA256}\\]"` which has `%{DATA:threat.indicator.type}` in all three `MD5` `SHA1` `SHA256` , then the pipeline adds the same value `file` to `threat.indicator.type` field thrice.

This started to happen when I started testing on 8.7.1-SNAPSHOT [ Possibly 8.7.0 might have it ] . I made a [workaround](https://github.com/elastic/integrations/pull/5910/files#diff-dae33f24a9ab60fd8007ceea3d2f5d92cfcff6821565c094708734c0fb929fb3L85-L89) for this in the pipeline to get through though.

**Comment by ruflin** [comment]
If we make these changes to the packages, is it mostly about the output data that slightly changes (improves?)? Will the same pipeline still work in older versions then 8.7 and users will have the same experience?

**Comment by HiDAl** [comment]
Hello, Y'all! 

@ishleenk17 has reached out asking about the change in the behavior of Grok's processor. We had issue #92092, which we treated as a bug and tackled in PR #92586. In summary, ES' Grok behaved differently than Logstash, so we decided to close the feature gap. 

thanks

**Comment by ishleenk17** [comment]
> Hello, Y'all!
> 
> @ishleenk17 has reached out asking about the change in the behavior of Grok's processor. We had issue #92092, which we treated as a bug and tackled in PR #92586. In summary, ES' Grok behaved differently than Logstash, so we decided to close the feature gap.
> 
> thanks

Would this change cause the behaviour as shown in the issue reported in this [comment](https://github.com/elastic/beats/issues/35133#issuecomment-1517317172)?
To me, it would look like duplication from user's perspective.

**Comment by bhapas** [comment]
@HiDAl Just a quick check , Should the Grok Processor consider output of duplicates? Like in the example you mentioned in https://github.com/elastic/elasticsearch/pull/92586/ 
```
filter {
    grok {
        match => { "message" => "^%{IPORHOST:user.address} (%{IPORHOST:user.address} )?" }
    }
}
```
And the event `"original": "{\"message\": \"81.12.23.34 81.12.23.34\"}"` - Same IP addresses in the event

So, Should the Grok Processor still output 

```
"user.address": [
    "81.12.23.34",
    "81.12.23.34"
  ],
```
or
`"user.address": "81.12.23.34"`

**Comment by bhapas** [comment]
> If we make **_these_** changes to the packages, is it mostly about the output data that slightly changes (improves?)? Will the same pipeline still work in older versions then 8.7 and users will have the same experience?


@ruflin 
**_these_** -- If you were mentioning the [workaround](https://github.com/elastic/integrations/pull/5910/files#diff-dae33f24a9ab60fd8007ceea3d2f5d92cfcff6821565c094708734c0fb929fb3L85-L89) , then yes it improves the output data to not show duplicates from 8.7
And yes , these changes should also work in older versions than 8.7 - I have currently (pre)fixed in packages running on older versions - currently `8.0.0`

**Comment by HiDAl** [comment]
@bhapas the output would include duplicates the same way that Logstash does. I've made the following test with Logstash and ES just to double-check they are doing the same:


## Logstash local test
```bash
> cat logstash_grok_sample.conf
...
filter {
    grok {
        match => { "message" => "^%{IPORHOST:useraddress} (%{IPORHOST:useraddress} )?%{GREEDYDATA:rest}" }
    }
}
...

# Add a new line to the log
> echo '{"message":"127.0.2.1 127.0.2.1 sdfasdfasdfasdf"}' >> sample-log.log
> tail -n1 sample-out.log | jq
{
  "event": {
    "original": "{\"message\":\"127.0.2.1 127.0.2.1 sdfasdfasdfasdf\"}"
  },
  "log": {
    "file": {
      "path": "/Users/palcantar/workspace/ES-scripts/sample-log.log"
    }
  },
  "@version": "1",
  "message": "127.0.2.1 127.0.2.1 sdfasdfasdfasdf",
  "@timestamp": "2023-04-21T08:19:25.124404Z",
  "useraddress": [
    "127.0.2.1",
    "127.0.2.1"
  ],
  "rest": "sdfasdfasdfasdf",
  "host": {
    "name": "Pablos-MBP.home"
  }
}

> echo '{"message":"127.0.2.1 sdfasdfasdfasdf"}' >> sample-log.log
> tail -n1 sample-out.log | jq
{
  "event": {
    "original": "{\"message\":\"127.0.2.1 sdfasdfasdfasdf\"}"
  },
  "log": {
    "file": {
      "path": "/Users/palcantar/workspace/ES-scripts/sample-log.log"
    }
  },
  "@version": "1",
  "message": "127.0.2.1 sdfasdfasdfasdf",
  "@timestamp": "2023-04-21T08:22:27.258881Z",
  "useraddress": "127.0.2.1",
  "rest": "sdfasdfasdfasdf",
  "host": {
    "name": "Pablos-MBP.home"
  }
}

> echo '{"message":"127.0.2.1 192.168.0.1 sdfasdfasdfasdf"}' >> sample-log.log 
> tail -n1 sample-out.log | jq
{
  "event": {
    "original": "{\"message\":\"127.0.2.1 192.168.0.1 sdfasdfasdfasdf\"}"
  },
  "log": {
    "file": {
      "path": "/Users/palcantar/workspace/ES-scripts/sample-log.log"
    }
  },
  "@version": "1",
  "message": "127.0.2.1 192.168.0.1 sdfasdfasdfasdf",
  "@timestamp": "2023-04-21T08:22:53.442535Z",
  "useraddress": [
    "127.0.2.1",
    "192.168.0.1"
  ],
  "rest": "sdfasdfasdfasdf",
  "host": {
    "name": "Pablos-MBP.home"
  }
}
```

## ES localtest
```json
POST /_ingest/pipeline/_simulate
{
  "pipeline": {
    "description": "_description",
    "processors": [
      {
        "grok": {
          "field": "message_str",
          "patterns": [
            "^%{IPORHOST:useraddress} (%{IPORHOST:useraddress} )?%{GREEDYDATA:rest}"
          ]
        }
      }
    ]
  },
  "docs": [
    {
      "_index": "index",
      "_id": "id",
      "_source": {
        "message_str": "127.0.2.1 127.0.2.1 sdfasdfasdfasdf"
      }
    }
  ]
}

{
  "docs": [
    {
      "doc": {
        "_index": "index",
        "_id": "id",
        "_version": "-3",
        "_source": {
          "rest": "sdfasdfasdfasdf",
          "useraddress": [
            "127.0.2.1",
            "127.0.2.1"
          ],
          "message_str": "127.0.2.1 127.0.2.1 sdfasdfasdfasdf"
        },
        "_ingest": {
          "timestamp": "2023-04-21T08:30:22.679096Z"
        }
      }
    }
  ]
}


POST /_ingest/pipeline/_simulate
{
  <<SAME PIPELINE>>
  },
  "docs": [
    {
      "_index": "index",
      "_id": "id",
      "_source": {
        "message_str": "127.0.2.1 sdfasdfasdfasdf"
      }
    }
  ]
}

{
  "docs": [
    {
      "doc": {
        "_index": "index",
        "_id": "id",
        "_version": "-3",
        "_source": {
          "rest": "sdfasdfasdfasdf",
          "useraddress": "127.0.2.1",
          "message_str": "127.0.2.1 sdfasdfasdfasdf"
        },
        "_ingest": {
          "timestamp": "2023-04-21T08:30:54.497687Z"
        }
      }
    }
  ]
}


POST /_ingest/pipeline/_simulate
{
  "pipeline": {
    <<SAME PIPELINE>>
  },
  "docs": [
    {
      "_index": "index",
      "_id": "id",
      "_source": {
        "message_str": "127.0.2.1 192.168.0.1 sdfasdfasdfasdf"
      }
    }
  ]
}

{
  "docs": [
    {
      "doc": {
        "_index": "index",
        "_id": "id",
        "_version": "-3",
        "_source": {
          "rest": "sdfasdfasdfasdf",
          "useraddress": [
            "127.0.2.1",
            "192.168.0.1"
          ],
          "message_str": "127.0.2.1 192.168.0.1 sdfasdfasdfasdf"
        },
        "_ingest": {
          "timestamp": "2023-04-21T08:31:33.40485Z"
        }
      }
    }
  ]
}
```

**Comment by bhapas** [comment]
> > > If we make **_these_** changes to the packages, is it mostly about the output data that slightly changes (improves?)? Will the same pipeline still work in older versions then 8.7 and users will have the same experience?
> > 
> > 
> > @ruflin **_these_** -- If you were mentioning the [workaround](https://github.com/elastic/integrations/pull/5910/files#diff-dae33f24a9ab60fd8007ceea3d2f5d92cfcff6821565c094708734c0fb929fb3L85-L89) , then yes it improves the output data to not show duplicates from 8.7 And yes , these changes should also work in older versions than 8.7 - I have currently (pre)fixed in packages running on older versions - currently `8.0.0`
> 
> @bhapas : The workaround will resolve the issue for now. But we should be mindful that such pipelines will be written in future also. So it would be good to have this resolved in ES. (if it is an issue et all).

@ishleenk17 Yes. And so it is a **workaround** . As @HiDAl mentioned ES have aligned the behaviour similar to Logstash , which might mean that integration/beats pipelines have to align accordingly? I am not sure though.

**Comment by klacabane** [comment]
Regarding multiline log entries in the elasticsearch slowlogs, they are possible when [index.indexing.slowlog.reformat](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules-slowlog.html) option is set to false and will log the source object as-is, newlines included. 
This became an issue when we [added support for structured logs](https://github.com/elastic/beats/pull/10447) because filebeat now interprets lines that starts with `{` as json log entries and we had to break support of this specific format (leading newlines, see [this comment for additional context](https://github.com/elastic/beats/pull/10447#discussion_r252512969)). [This elasticsearch change](https://github.com/elastic/elasticsearch/pull/38081) ensures that the source objects, while still able to span multiple lines, never start on a newline.

I'll update the test files to reflect the format change introduced in 7.0. We could add a guard for the unsupported format (leading newline) to prevent errors but I think we should let the pipeline throw so it gives user an opportunity to update the `reformat` option. @ruflin wdyt ?

**Comment by ishleenk17** [comment]
> @bhapas the output will include duplicates the same way that Logstash does. I've made the following test with Logstash and ES just to double-check they are doing the same:

Considering this to be a legitimate grom processing, I will be updating the golden files of MYSQL and IIS accordingly.

**Comment by ishleenk17** [comment]
@HiDAl : Is the change backward compatible ?

**Comment by ruflin** [comment]
@klacabane Trying to get my head around this. Does https://github.com/elastic/beats/pull/35164 break any previous versions of Elasticsearch logs?

**Comment by klacabane** [comment]
@ruflin Support for a specific format of slowlog plaintext entries (where a json source object starts on a newline) [was removed in 7.0](https://github.com/elastic/beats/pull/10447#discussion_r252512969). At the same time ES was updated to not output this format anymore. https://github.com/elastic/beats/pull/35164 just removes the unsupported format from the tests

**Comment by HiDAl** [comment]
Hi @ishleenk17, no, it's not. These pipeline templates aren't versioned? I mean, are they the same for 8.6, 8.7 or 7.x??

**Comment by ishleenk17** [comment]
> Hi @ishleenk17, no, it's not. These pipeline templates aren't versioned? I mean, are they the same for 8.6, 8.7 or 7.x??

There is no versioning in integrations. So this would be an issue in Integrations. We will either have to backport this fix or we would have to upgrade the kibana version for IIS/MYSQL. 

In beats, the changes go along with the releases so this won't be an issue.

**Comment by jsoriano** [comment]
> There is no versioning in integrations. So this would be an issue in Integrations. We will either have to backport this fix or we would have to upgrade the kibana version for IIS/MYSQL.

Ideally, the change done in the pipelines should be backwards compatible. This will probably imply to add a processor that converts the arrays with repeated values in single values.

**Comment by ishleenk17** [comment]
> > There is no versioning in integrations. So this would be an issue in Integrations. We will either have to backport this fix or we would have to upgrade the kibana version for IIS/MYSQL.
> 
> Ideally, the change done in the pipelines should be backwards compatible. This will probably imply to add a processor that converts the arrays with repeated values in single values.

I have raised a new [issue](https://github.com/elastic/integrations/issues/6016) in Integrations to discuss the appropriate fix. 
Please feel free to share your input over there.
